jsettlers.logic/src/main/java/jsettlers/logic/movable/components/Component.java

```diff
@@ -0,0 +1,119 @@
+package jsettlers.logic.movable.components;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+
+import jsettlers.logic.movable.Entity;
+import jsettlers.logic.movable.Notification;
+
+public abstract class Component implements Serializable {
+    private static final long serialVersionUID = -3071296154652495126L;
+    public Entity entity;
+    private HashSet<Notification> consumedNotifications = new HashSet<>();
+
+    /**
+     * Called once when the entity gets enabled for the first time
+     */
+    public final void Awake() {
+        onAwake();
+    }
+    protected void onAwake() {}
+
+    public final void Update() {
+        consumedNotifications.clear();
+        onUpdate();
+    }
+    protected void onUpdate() {}
+
+    public final void LateUpdate() { onLateUpdate(); }
+    protected void onLateUpdate() {}
+
+    /**
+     *  Called when the entity is set to active
+     */
+    public final void Enable() { onEnable(); }
+    protected void onEnable() {}
+
+    /**
+     *  Called when the entity is set to inactive
+     */
+    public final void Disable() { onDisable(); }
+    protected void onDisable() {}
+
+    /**
+     *  Called before the entity gets destroyed/killed
+     */
+    public final void Destroy() { onDestroy(); }
+    protected void onDestroy() {}
+
+    public <T extends Notification> List<T> getNotifications(Class<T> type) {
+        Iterator<T> it = getNotificationsIt(type);
+        List<T> result = new ArrayList<T>();
+        while (it.hasNext()) {
+            result.add(it.next());
+        }
+        return result;
+    }
+
+    public <T extends Notification> T getNextNotification(Class<T> type, boolean consume) {
+        Iterator<T> it = getNotificationsIt(type);
+        T note = it.next();
+        if (note != null) {
+            if (consume) consumeNotification(note);
+        }
+        return null;
+    }
+
+    public <T extends Notification> boolean containsNotification(Class<T> type) {
+        return getNotificationsIt(type).hasNext();
+    }
+
+    public <T extends Notification> Iterator<T> getNotificationsIt(Class<T> type) {
```

Comment:
Some things here:

* I think it would be better to realize this with a java8 stream. From what I found, you always only check if there is a single element and then get that element. You could make this function return a Stream<T> and build the stream like this:

```
   return stream(entity.getAllNotifications()).filter(type::isInstance).map(notification -> (T)notification);
```

and when you want to get the first matching notification just do

```
getNotificationsIt(type).findAny().isPresent()
```

Furthermore, I would then rename this method to something like getNotificationsOfType.

* Regarding the consumed status of notifcations. Have you thought about adding a consumed flag to the Notification parent class along with a `consume()` method? This way you can save this state of the notification in the notification, where it belongs (more OOP). (and it should be better in performance compared to the hash set lookup).

* Is there a need to use `type.isInstance()`? I mean will there be the use case that you ask for a supertype notification and expect to get a child class of it? If not, we could simply do a == comparision on the class of the notification, which should be considerably faster.

Reply:
> Regarding the consumed status of notifcations.
It is important that each component is able to track which notifications got consumed so far individually. Right now the `GetNotificationsOfType` returns the objects from the entity. This ensures the equality of `Notifications` across `Component` borders. If the consumed flag is part of the `Notification` itself, then we have to override the standard equality of Notifications.